smart-contracts/contracts/mixins/MixinRefunds.sol

```diff
@@ -39,19 +44,40 @@ contract MixinRefunds is
   function cancelAndRefund()
     external
   {
-    Key storage key = _getKeyFor(msg.sender);
+    _cancelAndRefund(msg.sender);
+  }
 
-    uint refund = _getCancelAndRefundValue(msg.sender);
+  /**
+   * @dev Cancels a key owned by a different user and sends the funds to the msg.sender.
+   * @param _keyOwner this user's key will be canceled
+   * @param _signature getCancelAndRefundApprovalHash signed by the _keyOwner
+   */
+  function cancelAndRefundFor(
+    address _keyOwner,
+    bytes calldata _signature
+  ) external
+  {
+    require(
+      ECDSA.recover(
+        ECDSA.toEthSignedMessageHash(
+          getCancelAndRefundApprovalHash(_keyOwner, msg.sender)
+        ),
+        _signature
+      ) == _keyOwner, 'INVALID_SIGNATURE'
+    );
 
-    emit CancelKey(key.tokenId, msg.sender, refund);
-    // expirationTimestamp is a proxy for hasKey, setting this to `block.timestamp` instead
-    // of 0 so that we can still differentiate hasKey from hasValidKey.
-    key.expirationTimestamp = block.timestamp;
+    keyOwnerToNonce[_keyOwner]++;
+    _cancelAndRefund(_keyOwner);
+  }
 
-    if (refund > 0) {
-      // Security: doing this last to avoid re-entrancy concerns
-      _transfer(msg.sender, refund);
-    }
+  /**
+   * @dev Increments the current nonce for the msg.sender.
+   * This can be used to invalidate a previously signed message.
+   */
+  function incrementNonce(
```

Comment:
If this function is called by the _keyOwner after they've signed/approved a cancellation, but before `cancelAndRefundFor` is called, they can prevent it from being successful because `getCancelAndRefundApprovalHash` will no longer return a the right hash (because of the nonce change).

Reply:
Correct.  However a more effective attack would be to simply call `cancelAndRefund` before we do.  I added a comment RE this in the test.  The backend implementation can guard against this by waiting for the tx to be mined before issuing the CC refund - or we accept this risk.

Reply:
Yeah , I think waiting for the tx to be mined is worthwhile.